### DISCUSSION TITLE
How to avoid duplicated array entry in machine config patches?

### DISCUSSION THREAD

**Discussion by stevefan1999-personal** [siderolabs/talos#8177] [Q&A]
Hi, I'm using a GitOps based workflow with patch.yaml file, and so I find this behavior quite counterintuitive:
```diff
diff --git a/patch.yaml b/patch.yaml
index 08a65d0..768befd 100644
--- a/patch.yaml
+++ b/patch.yaml
@@ -1,6 +1,7 @@
 machine:
   kubelet:
     extraArgs:
+      cloud-provider: external
       rotate-server-certificates: true
     nodeIP:
       validSubnets:
@@ -18,7 +19,12 @@ machine:
     kubePrism:
       enabled: true
       port: 7445
-
+    kubernetesTalosAPIAccess:
+      enabled: true
+      allowedRoles:
+      - os:reader
+      allowedKubernetesNamespaces:
+      - kube-system
 cluster:
   discovery:
     enabled: true
@@ -39,3 +45,7 @@ cluster:
   extraManifests:
   - https://raw.githubusercontent.com/alex1989hu/kubelet-serving-cert-approver/main/deploy/standalone-install.yaml
   - https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml
+  externalCloudProvider:
+    enabled: true
+    manifests:
+    - https://raw.githubusercontent.com/siderolabs/talos-cloud-controller-manager/main/docs/deploy/cloud-controller-manager.yml

```

So I just added some entries in the patch.yaml and I want to refresh the existing controlplane.yaml and worker.yaml for fast node bootstrap access. So I run `talosctl machineconfig patch controlplane.yaml --patch "@patch.yaml" -o controlplane.yaml`, but I noticed something is off:
```yaml
    extraManifests:
        - https://raw.githubusercontent.com/alex1989hu/kubelet-serving-cert-approver/main/deploy/standalone-install.yaml
        - https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml
        - https://raw.githubusercontent.com/alex1989hu/kubelet-serving-cert-approver/main/deploy/standalone-install.yaml
        - https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml
```
As you can see, those manifests are duplicated, when they intuitively shouldn't, because from a human standpoint, the original array and the new array is naturally the same, so there's no need to intervene it, but from the machine standpoint, or rather the current array merge strategy in talosctl, is to just append the array by default. This would still happen if I output this to a new file, and then thoroughly replacing the old with the new by forcefully `mv`ing it. So there is currently no workaround regarding this unless I use JSON patch.

I don't want to bother with JSON patches, because they are quite a mess to work with, plus I have Git to manage all the config differences, so I would like to use the patch.yaml as an overlay of the original files (or should I say over the files that are currently committed). I rather want to see if I can have retained the original array unless there really are some differences. 

So what I, a normal human would think of this situation, is to treat the array as a hash set, and compare if there are set differences from the patch and the existing entries (heck this may even look like it comes from a machine mindset).

**Comment by smira** [comment]
You can use other tools to manage YAML files, but the supported Talos Linux flow for managing machine configuration assumes that a set of patches is applied to the base configuration. Patching with the same patch is not always idempotent.

* [Recommended flow](https://www.talos.dev/v1.6/introduction/prodnotes/#configure-talos)
* [Merge semantics](https://www.talos.dev/v1.6/talos-guides/configuration/patching/#strategic-merge-patches)

**Reply by stevefan1999-personal** [reply]
That's why I said it is not very intuitive to typical users

**Comment by nightguide** [comment]
@smira hi!

I’d like to return to this discussion and I fully support @stevefan1999-personal on this matter.

The strategic merge should follow the classic `replace` behavior for list-type elements, rather than attempting to append to existing lists when a patch is applied multiple times.

I’d also like to note that **starting from Talos v1.12.0, a multi-document configuration format will be used, and JSON patches will no longer be applicable since** - _“JSON patches don’t support multi-document machine configuration, while strategic merge patches do.”_

Are there any plans to deprecate the JSON Patch functionality in the near future?
Will the strategic merge flow be changed to use `replace` for list elements?

Thank you.

**Comment by nightguide** [comment]
@smira @stevefan1999-personal Hi!

I’ve opened an issue — see https://github.com/siderolabs/talos/issues/12005
